btf/handle.go

```diff
@@ -119,3 +119,33 @@ func (it *HandleIterator) Next(handle **Handle) bool {
 func (it *HandleIterator) Err() error {
 	return it.err
 }
+
+// FindHandle returns the first handle for which predicate returns true.
+//
+// Requires CAP_SYS_ADMIN.
+//
+// Returns ErrNotFound if predicate never returns true or if there is no BTF
+// loaded into the kernel.
+func FindHandle(predicate func(info *HandleInfo) bool) (*Handle, error) {
+	var handle *Handle
+	defer handle.Close()
+
+	it := new(HandleIterator)
+	for it.Next(&handle) {
+		info, err := handle.Info()
+		if err != nil {
+			return nil, fmt.Errorf("info for ID %d: %w", it.ID, err)
+		}
+
+		if predicate(info) {
+			tmp := handle
```

Comment:
The point is to avoid `defer handle.Close()` closing the handle we are returning. It is used in `findTargetInModule` as well.

Not sure how to solve this without adding the same comment to every callsite. Maybe I need to change HandleIterator to own the handle after all.

Reply:
Ah, I see. That is a cool pattern. Changing ownership is difficult since you are returning the handle while the iterator goes out of scope.

Any reason for not using `runtime.SetFinalizer` like with other objects that hold file descriptors like maps and programs?

Reply:
We already set a finalizer, but it's better to explicitly Close since the former is run from the GC. There is no guarantee how often the GC runs, so in the meantime leaked handles can use up fds. This in turn leads to fd exhaustion.

I've changed HandleIterator slighty and added a `Take` method which should make this trick more obvious: https://github.com/cilium/ebpf/pull/737/commits/2b8def66cc76a75d1a5643b632d2175b668b8e7c